### PR TITLE
dnsdist-1.8.x: Properly handle short reads on backend upgrade discovery

### PR DESCRIPTION
**Pull request by rgacogne**
### Short description

Backport of #13113 to rel/dnsdist-1.8.x

### Checklist

I have:
- [x] read the [CONTRIBUTING.md](https://github.com/PowerDNS/pdns/blob/master/CONTRIBUTING.md) document
- [x] compiled this code
- [ ] tested this code
- [ ] included documentation (including possible behaviour changes)
- [x] documented the code
- [x] added or modified regression test(s)
- [ ] added or modified unit test(s)
- [x]  checked that this code was merged to master

### CODE REVIEW

File: pdns/dnsdistdist/dnsdist-discovery.cc [hunk 1]
```diff
@@ -272,8 +272,9 @@ bool ServiceDiscovery::getDiscoveredConfig(const UpgradeableBackend& upgradeable
 
     sock.writenWithTimeout(reinterpret_cast<const char*>(packet.data()), packet.size(), backend->d_config.tcpSendTimeout);
 
+    const struct timeval remainingTime = {.tv_sec = backend->d_config.tcpRecvTimeout, .tv_usec = 0};
     uint16_t responseSize = 0;
-    auto got = sock.readWithTimeout(reinterpret_cast<char*>(&responseSize), sizeof(responseSize), backend->d_config.tcpRecvTimeout);
+    auto got = readn2WithTimeout(sock.getHandle(), &responseSize, sizeof(responseSize), remainingTime);
     if (got < 0 || static_cast<size_t>(got) != sizeof(responseSize)) {
```
Comment (omoerbeek) [inline pdns/dnsdistdist/dnsdist-discovery.cc:278]:
```suggestion
    if (got < 0 || static_cast<size_t>(got) != sizeof(responseSize)) {
```
as `readn2WithTimeout` returns  `size_t` this test is redundant.

Comment (Habbie) [reply pdns/dnsdistdist/dnsdist-discovery.cc:278]: #13138 :)

Comment (rgacogne) [reply pdns/dnsdistdist/dnsdist-discovery.cc:278]: Right, I'll include the commits from the second follow-up PR in this one :)

Comment (rgacogne) [reply pdns/dnsdistdist/dnsdist-discovery.cc]: Done!

**Pull request merged**